packages/babel-plugin-transform-typescript/src/enum.js

```diff
@@ -0,0 +1,210 @@
+      if (seen(path.parentPath.parentPath)) {
+        path.remove();
+      } else {
+        path.replaceWith(makeVar(node.id, t, "let"));
+      }
+      break;
+    }
+
+    default:
+      throw new Error(`Unexpected enum parent '${path.parent.type}`);
+  }
+
+  function seen(parentPath: Path<Node>) {
+    if (parentPath.getData(name)) {
+      return true;
+    } else {
+      parentPath.setData(name, true);
+      return false;
+    }
+  }
+}
+
+function makeVar(id, t, kind): VariableDeclaration {
+  return t.variableDeclaration(kind, [t.variableDeclarator(id)]);
+}
+
+/**
+ * Generates the statement that fills in the variable declared by the enum.
+ * `(function (E) { ... assignments ... })(E || (E = {}));`
+ */
+function enumFill(path, t, id) {
+  const x = translateEnumValues(path, t);
+  const assignments = x.map(([memberName, memberValue]) => {
+    const inner = t.assignmentExpression(
+      "=",
+      t.memberExpression(id, t.stringLiteral(memberName), /*computed*/ true),
+      memberValue,
+    );
+    const outer = t.assignmentExpression(
+      "=",
+      t.memberExpression(id, inner, /*computed*/ true),
+      t.stringLiteral(memberName),
+    );
+    return t.expressionStatement(outer);
+  });
+
+  // E || (E = {})
+  const callArg = t.logicalExpression(
+    "||",
+    id,
+    t.assignmentExpression("=", id, t.objectExpression([])),
+  );
+  const body = t.blockStatement(assignments);
+  const callee = t.functionExpression(null, [id], body);
+  return t.expressionStatement(t.callExpression(callee, [callArg]));
+}
+
+/**
+ * Maps the name of an enum member to its value.
+ * We keep track of the previous enum members so you can write code like:
+ *   enum E {
+ *     X = 1 << 0,
+ *     Y = 1 << 1,
+ *     Z = X | Y,
+ *   }
+ */
+type PreviousEnumMembers = { [name: string]: number | typeof undefined };
+
+function getName(memberName) {
+  return memberName.type === "Identifier" ? memberName.name : memberName.value;
+}
+
+function translateEnumValues(path, t) {
+  const seen: PreviousEnumMembers = Object.create(null);
+  // Start at -1 so the first enum member is its increment, 0.
+  let prev: number | typeof undefined = -1;
+  return path.node.members.map(member => {
+    const name = getName(member.id);
+    const initializer = member.initializer;
+    let value: Expression;
+    if (initializer) {
+      const constValue = evaluate(initializer, seen);
+      if (constValue !== undefined) {
+        value = t.numericLiteral(constValue);
+        prev = constValue;
+      } else {
+        value = initializer;
+        prev = undefined;
+      }
+    } else {
+      if (prev !== undefined) {
+        prev++;
+        value = t.numericLiteral(prev);
+      } else {
+        throw path.buildCodeFrameError("Enum member must have initializer.");
+      }
+    }
+
+    return [name, value];
+  });
+}
+
+// Based on the TypeScript repository's `evalConstant` in `checker.ts`.
+function evaluate(expr, seen: PreviousEnumMembers) {
```

Comment:
Also, we need to be able to evaluate previously-seen enum values, so any function on just the path wouldn't be enough:
```ts
enum E {
  A = 1,
  B = A * 2,
  C, // inferred as 3
}
```